[nonisolated-nonsending] Make the AST not consider nonisolated(nonsending) to be an actor isolation crossing point.

gottesmm opened the pull request:
We were effectively working around this previously at the SIL level. This caused
us not to obey the semantics of the actual evolution proposal. As an example of
this, in the following, x should not be considered main actor isolated:

```swift
nonisolated(nonsending) func useValue<T>(_ t: T) async {}
@MainActor func test() async {
  let x = NS()
  await useValue(x)
  print(x)
}
```

we should just consider this to be a merge and since useValue does not have any
MainActor isolated parameters, x should not be main actor isolated and we should
not emit an error here.

I also fixed a separate issue where we were allowing for parameters of
nonisolated(nonsending) functions to be passed to `@concurrent` functions. We
cannot allow for this to happen since the nonisolated(nonsending) parameters
/could/ be actor isolated. Of course, we have lost that static information at
this point so we cannot allow for it. Given that we have the actual dynamic
actor isolation information, we could dynamically allow for the parameters to be
passed... but that is something that is speculative and is definitely outside of
the scope of this patch.

rdar://154139237

gottesmm committed:
[nonisolated-nonsending] Make the AST not consider nonisolated(nonsending) to be an actor isolation crossing point.

We were effectively working around this previously at the SIL level. This caused
us not to obey the semantics of the actual evolution proposal. As an example of
this, in the following, x should not be considered main actor isolated:

```swift
nonisolated(nonsending) func useValue<T>(_ t: T) async {}
@MainActor func test() async {
  let x = NS()
  await useValue(x)
  print(x)
}
```

we should just consider this to be a merge and since useValue does not have any
MainActor isolated parameters, x should not be main actor isolated and we should
not emit an error here.

I also fixed a separate issue where we were allowing for parameters of
nonisolated(nonsending) functions to be passed to @concurrent functions. We
cannot allow for this to happen since the nonisolated(nonsending) parameters
/could/ be actor isolated. Of course, we have lost that static information at
this point so we cannot allow for it. Given that we have the actual dynamic
actor isolation information, we could dynamically allow for the parameters to be
passed... but that is something that is speculative and is definitely outside of
the scope of this patch.

rdar://154139237

gottesmm committed:
[rbi] Convert all rbi tests to run also in NonisolatedNonsendingByDefault.

Going to update the tests in the next commit. This just makes it easier to
review.

xedin commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -8006,13 +8023,21 @@ ActorReferenceResult ActorReferenceResult::forReference(
+    // caller isolation inheriting. Check if it is nonisolated. In such a case,
+    // if our context isolation is caller isolation inheriting, we want to emit
+    // an error since parameters that are caller isolation inheriting cannot be
+    if (decl->isAsync() && contextIsolation.isCallerIsolationInheriting())
```
I think we should refactor this whole block into:

```
if (declRef.getDecl()
             ->getAttrs()
             .hasAttribute<UnsafeInheritExecutorAttr>())
  return forSameConcurrencyDomain(declIsolation, options);

if (decl->isAsync() && (contextIsolation.isActorIsolated() || contextIsolation.isCallerIsolationInheriting()))
  return forExitsActorToNonisolated(contextIsolation, options);

// Otherwise, we stay in the same concurrency domain, whether on an actor
// or in a task.
return forSameConcurrencyDomain(declIsolation, options);
```

**But** there is also a bit of a caveat in here re: `contextIsolation.isCallerIsolationInheriting()`: what if the caller of "context" is also nonisolated, this means that the reference would not actually cross the isolation boundary doesn't it?

gottesmm replied:
Yes. The problem is that statically we do not have that information. So we need to treat it conservatively. That is why I am in a separate PR going to add an educational note with that information. I am just workshopping what it looks like and it shouldn't hold up this PR.

gottesmm committed:
[rbi] Wrap use ActorIsolation::printForDiagnostics with our own SILIsolationInfo::printActorIsolationForDiagnostics.

I am doing this so that I can change how we emit the diagnostics just for
SendNonSendable depending on if NonisolatedNonsendingByDefault is enabled
without touching the rest of the compiler.

This does not actually change any of the actual output though.

gottesmm committed:
[rbi] Begin tracking if a function argument is from a nonisolated(nonsending) parameter and adjust printing as appropriate.

Specifically in terms of printing, if NonisolatedNonsendingByDefault is enabled,
we print out things as nonisolated/task-isolated and @concurrent/@concurrent
task-isolated. If said feature is disabled, we print out things as
nonisolated(nonsending)/nonisolated(nonsending) task-isolated and
nonisolated/task-isolated. This ensures in the default case, diagnostics do not
change and we always print out things to match the expected meaning of
nonisolated depending on the mode.

I also updated the tests as appropriate/added some more tests/added to the
SendNonSendable education notes information about this.

gottesmm committed:
[rbi] Thread through a SILFunction into print routines so we can access LangOpts.Features so we can change how we print based off of NonisolatedNonsendingByDefault.

We do not actually use this information yet though... This is just to ease
review.

xedin reviewed (approved):
LGTM! We should be careful about using StringRef in the diagnostic though, the arguments need to be owned by the ASTContext to make that the diagnostic doesn't outlive the argument.

gottesmm committed:
[rbi] Use interned StringRefs for diagnostics instead of SmallString<64>.

This makes the code easier to write and also prevents any lifetime issues from a
diagnostic outliving the SmallString due to diagnostic transactions.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
Failed in lldb

gottesmm commented:
@swift-ci smoke test macOS platform
